src/server.rs

```diff
@@ -140,15 +141,61 @@ pub fn listen<S, Req, Resp, E>(new_service: S,
 pub struct Handle {
     reactor: reactor::Core,
     addr: SocketAddr,
+    shutdown: Shutdown,
+}
+
+/// A hook to shut down a running server.
+#[derive(Clone)]
+pub struct Shutdown {
+    tx: mpsc::UnboundedSender<()>,
+}
+
+impl Shutdown {
+    /// Signals to the server to enter lame duck mode: existing connections will be served,
+    /// but no new connections will be accepted.
+    pub fn shutdown(self) {
+        if let Err(e) = self.tx.send(()) {
+            info!("Failed to shutdown server: {}", e);
+        }
+    }
 }
 
 impl Handle {
     #[doc(hidden)]
-    pub fn new(reactor: reactor::Core, addr: SocketAddr) -> Self {
-        Handle {
-            reactor: reactor,
-            addr: addr,
-        }
+    pub fn listen<S, Req, Resp, E>(new_service: S,
+                                   addr: SocketAddr,
+                                   options: Options)
+                                   -> io::Result<Self>
+        where S: NewService<Request = Result<Req, bincode::Error>,
+                            Response = Response<Resp, E>,
+                            Error = io::Error> + 'static,
+              Req: Deserialize + 'static,
+              Resp: Serialize + 'static,
+              E: Serialize + 'static
+    {
+        let reactor = reactor::Core::new()?;
+        let (addr, server) = listen(new_service, addr, &reactor.handle(), options)?;
+        let (tx, rx) = mpsc::unbounded();
+        let shutdown = rx.into_future()
+            .map_err(|_| warn!("UnboundedReceiver resolved to an Err; can it do that?"))
+            .and_then(|(result, _)| {
+                match result {
+                    Some(()) => {
+                        debug!("Got shutdown request.");
+                        future::Either::A(future::ok(()))
+                    }
+                    None => {
+                        debug!("Shutdown hook dropped; never shutting down.");
+                        future::Either::B(future::empty())
+                    }
+                }
+            });
+        reactor.handle().spawn(server.select(shutdown).then(|_| {
+            debug!("Entering lame duck mode.");
```

Comment:
It means existing connections will be honored but no new connections will be accepted.

Yeah, so I was a bit hesitant to implement that as part of this, because it'd require one of the following, all of which I wanted to spend a bit more time thinking about and getting feedback on:

1. Have the `Handle` hold the server future so that it can run it directly. Then `run` return once the server stops accepting connections. This requires exposing some type parameters on `Handle`.
2. Same as 1 but box the server to avoid type parameters. Probably not a huge performance hit since, but would be better to benchmark first.
3. Signal shutdown some other way, maybe via an `AtomicBool`. Not sure if there'd be any overhead in checking an atomic once per iteration of `loop { reactor.turn(None); }`.